In [2]:
# imports to look in "leap-persistent" google cloud storage
import gcsfs
fs = gcsfs.GCSFileSystem()

# looking in a cloud storage directory

### To look at contents in a directory, use '.ls' and put your path in the parentheses. 

### The path will always start with " gs://leap-persistent/ "

We insert your username and a path associated with the default runs

In [54]:
your_username = 'azamkhan99' 
fs.ls(f'gs://leap-persistent/{your_username}/{your_username}/')

['leap-persistent/azamkhan99/azamkhan99/pco2_residual']

In [31]:
accesmd_df = xr.open_zarr('gs://leap-persistent/azamkhan99/azamkhan99/pco2_residual/nmse/post02_xgb/reconstructions/ACCESS-ESM1-5/member_r10i1p1f1/recon_pCO2residual_ACCESS-ESM1-5_member_r10i1p1f1_mon_1x1_200401_202312.zarr').to_dataframe()

In [28]:
accesmd_df

pCO2_recon_full  pCO2_recon_unseen  pCO2_truth
time       ylat  xlon                                                  
2004-01-01 -89.5 -179.5              NaN                NaN         NaN
                 -178.5              NaN                NaN         NaN
                 -177.5              NaN                NaN         NaN
                 -176.5              NaN                NaN         NaN
                 -175.5              NaN                NaN         NaN
...                                  ...                ...         ...
2023-12-01  89.5  175.5              NaN                NaN  349.393516
                  176.5              NaN                NaN  349.338248
                  177.5              NaN                NaN  349.282905
                  178.5              NaN                NaN  349.227450
                  179.5              NaN                NaN  349.171919

[15552000 rows x 3 columns]

In [13]:
import xarray as xr

In [15]:
canesm = xr.open_zarr('gs://'+'leap-persistent/azamkhan99/azamkhan99/pco2_residual/nmse/post02_xgb/reconstructions/CanESM5/member_r1i1p1f1/recon_pCO2residual_CanESM5_member_r1i1p1f1_mon_1x1_200401_202312.zarr')

In [18]:
canesm_df = canesm.to_dataframe()

In [24]:
canesm_df.dropna()

pCO2_recon_full  net_mask  pCO2_recon_full_mean  \
time       xlon   ylat                                                     
2004-01-01 -179.5 -76.5        -3.126162       1.0             -3.126162   
                  -75.5         3.231421       1.0              3.231421   
                  -74.5        -3.149131       1.0             -3.149131   
                  -73.5        -3.479035       1.0             -3.479035   
                  -72.5        -2.646990       1.0             -2.646990   
...                                  ...       ...                   ...   
2023-12-01  179.5  57.5        12.318372       1.0             12.318372   
                   58.5        10.298716       1.0             10.298716   
                   59.5         9.355980       1.0              9.355980   
                   60.5         9.353810       1.0              9.353810   
                   61.5        22.842762       1.0             22.842762   

                         pCO2_recon_unseen_mean  socat_mask  pCO2_truth  \
time       xlon   ylat                                                    
2004-01-01 -179.5 -76.5               -3.126162         0.0   -7.623447   
                  -75.5                3.231421         0.0   -5.219215   
                  -74.5               -3.149131         0.0   -3.870247   
                  -73.5               -3.479035         0.0   -4.868774   
                  -72.5               -2.646990         0.0   -2.944900   
...                                         ...         ...         ...   
2023-12-01  179.5  57.5               12.318372         0.0   10.277021   
                   58.5               10.298716         0.0    8.019634   
                   59.5                9.355980         0.0    6.964161   
                   60.5                9.353810         0.0   12.263339   
                   61.5               22.842762         0.0   32.152586   

                         pCO2_recon_unseen_std  pCO2_recon_unseen  \
time       xlon   ylat                                              
2004-01-01 -179.5 -76.5              10.209448          -3.126162   
                  -75.5               8.702923           3.231421   
                  -74.5               7.543845          -3.149131   
                  -73.5               6.608551          -3.479035   
                  -72.5               5.256913          -2.646990   
...                                        ...                ...   
2023-12-01  179.5  57.5               9.673994          12.318372   
                   58.5              10.050552          10.298716   
                   59.5              11.504381           9.355980   
                   60.5              11.530214           9.353810   
                   61.5              12.507779          22.842762   

                         pCO2_recon_full_std  
time       xlon   ylat                        
2004-01-01 -179.5 -76.5            10.209448  
                  -75.5             8.702923  
                  -74.5             7.543845  
                  -73.5             6.608551  
                  -72.5             5.256913  
...                                      ...  
2023-12-01  179.5  57.5             9.673994  
                   58.5            10.050552  
                   59.5            11.504381  
                   60.5            11.530214  
                   61.5            12.507779  

[7868928 rows x 9 columns]

# getting file sizes

### if we want to get info for a file on the cloud storage:

In [14]:
# a random csv!
fs.info('gs://leap-persistent/azamkhan99/test_folder/test.csv')

FileNotFoundError: leap-persistent/azamkhan99/test_folder

### if we want specifically the size, we use 'fs.du()' which stands for "disk usage":

In [5]:
fs.du('gs://leap-persistent/galenmckinley/test_folder/test.csv')

17

### just "17" is unclear though -> this function will make this clearer with units!

In [6]:
def convert_to_human_readable(size):
    """
    Converts disk usage size to 'human readable' format!
    """
    
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size < 1024:
            return f"{size:.2f} {unit}"
        size /= 1024

In [7]:
convert_to_human_readable(17)

'17.00 B'

### that csv is 17 bytes!

### here's another example, but with a .zarr file:

In [8]:
fs.du('gs://leap-persistent/galenmckinley/pco2_residual/nmse/post02_xgb/reconstructions/CanESM5/member_r1i1p1f1/recon_pCO2residual_CanESM5_member_r1i1p1f1_mon_1x1_200401_202312.zarr')

166750349

In [9]:
convert_to_human_readable(21858071)

'20.85 MB'

that .zarr file is 20.85 megabytes

# Deleting a file
To remove a file, you will use 'fs.rm(path)'

# ***Be very careful*** 

Make sure you are sure that you are removing what you want and that you are not deleting in someone else's directory.

In [11]:
# do not actually run this code, please!

# fs.rm('gs://leap-persistent/galenmckinley/test_folder/test.csv')

### Things are more complicated with .zarr files because .zarr files are considered "directories" by the file system
### If deleting a .zarr file, you will just need to add the argument 'recursive=True' inside the parentheses

In [12]:
# fs.rm('gs://leap-persistent/....../filename.zarr', recursive=True)